Bug when getting artifact version by content if artifact has multiple versions with identical content

```

2022-03-21 20:18:36 INFO <d5cca8a0-0fa3-4ff9-833e-dcc93448d68b> [io.apicurio.registry.metrics.health.liveness.ResponseErrorLivenessCheck] (executor-thread-532) After this event, the error counter is 3 (out of the maximum 5 allowed).
2022-03-21 20:18:36 ERROR <d5cca8a0-0fa3-4ff9-833e-dcc93448d68b> [io.apicurio.registry.services.http.RegistryExceptionMapperService] (executor-thread-532) [500 ERROR DETECTED] : io.apicurio.registry.storage.impl.sql.jdb.RuntimeSqlException: SQL error: Expected only one result but got multiple.: io.apicurio.registry.storage.RegistryStorageException: io.apicurio.registry.storage.impl.sql.jdb.RuntimeSqlException: SQL error: Expected only one result but got multiple.
	at io.apicurio.registry.storage.impl.sql.AbstractSqlRegistryStorage.getArtifactVersionMetaData(AbstractSqlRegistryStorage.java:1253)
	at io.apicurio.registry.storage.impl.sql.SqlRegistryStorage_Subclass.getArtifactVersionMetaData$$superforward1(SqlRegistryStorage_Subclass.zig:8488)
	at io.apicurio.registry.storage.impl.sql.SqlRegistryStorage_Subclass$$function$$41.apply(SqlRegistryStorage_Subclass$$function$$41.zig:55)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.apicurio.common.apps.logging.LoggingInterceptor.logMethodEntry(LoggingInterceptor.java:53)
	at io.apicurio.common.apps.logging.LoggingInterceptor_Bean.intercept(LoggingInterceptor_Bean.zig:327)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.apicurio.registry.metrics.health.readiness.PersistenceTimeoutReadinessInterceptor.intercept(PersistenceTimeoutReadinessInterceptor.java:32)
	at io.apicurio.registry.metrics.health.readiness.PersistenceTimeoutReadinessInterceptor_Bean.intercept(PersistenceTimeoutReadinessInterceptor_Bean.zig:429)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.apicurio.registry.metrics.health.liveness.PersistenceExceptionLivenessInterceptor.intercept(PersistenceExceptionLivenessInterceptor.java:27)
	at io.apicurio.registry.metrics.health.liveness.PersistenceExceptionLivenessInterceptor_Bean.intercept(PersistenceExceptionLivenessInterceptor_Bean.zig:378)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:132)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:103)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:57)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(TransactionalInterceptorRequired_Bean.zig:340)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.apicurio.registry.metrics.StorageMetricsInterceptor.intercept(StorageMetricsInterceptor.java:48)
	at io.apicurio.registry.metrics.StorageMetricsInterceptor_Bean.intercept(StorageMetricsInterceptor_Bean.zig:429)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.apicurio.registry.storage.impl.sql.SqlRegistryStorage_Subclass.getArtifactVersionMetaData(SqlRegistryStorage_Subclass.zig:16026)
	at io.apicurio.registry.storage.impl.sql.SqlRegistryStorage_ClientProxy.getArtifactVersionMetaData(SqlRegistryStorage_ClientProxy.zig:749)
	at io.apicurio.registry.storage.decorator.RegistryStorageDecorator.getArtifactVersionMetaData(RegistryStorageDecorator.java:341)
	at io.apicurio.registry.mt.limits.RegistryStorageLimitsEnforcer_ClientProxy.getArtifactVersionMetaData(RegistryStorageLimitsEnforcer_ClientProxy.zig:879)
	at io.apicurio.registry.storage.RegistryStorageProducer_ProducerMethod_realImpl_cf1c876861dd1c25dca504d30a12bfedeafd47bd_ClientProxy.getArtifactVersionMetaData(RegistryStorageProducer_ProducerMethod_realImpl_cf1c876861dd1c25dca504d30a12bfedeafd47bd_ClientProxy.zig:546)
	at io.apicurio.registry.ccompat.store.RegistryStorageFacadeImpl.getSchema(RegistryStorageFacadeImpl.java:156)
	at io.apicurio.registry.ccompat.store.RegistryStorageFacadeImpl_ClientProxy.getSchema(RegistryStorageFacadeImpl_ClientProxy.zig:525)
	at io.apicurio.registry.ccompat.rest.impl.SubjectsResourceImpl.findSchemaByContent(SubjectsResourceImpl.java:52)
	at io.apicurio.registry.ccompat.rest.impl.SubjectsResourceImpl_Subclass.findSchemaByContent$$superforward1(SubjectsResourceImpl_Subclass.zig:352)
	at io.apicurio.registry.ccompat.rest.impl.SubjectsResourceImpl_Subclass$$function$$1.apply(SubjectsResourceImpl_Subclass$$function$$1.zig:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.apicurio.common.apps.logging.LoggingInterceptor.logMethodEntry(LoggingInterceptor.java:53)
	at io.apicurio.common.apps.logging.LoggingInterceptor_Bean.intercept(LoggingInterceptor_Bean.zig:327)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.apicurio.registry.auth.AuthorizedInterceptor.authorizeMethod(AuthorizedInterceptor.java:149)
	at io.apicurio.registry.auth.AuthorizedInterceptor_Bean.intercept(AuthorizedInterceptor_Bean.zig:743)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.apicurio.registry.ccompat.rest.impl.SubjectsResourceImpl_Subclass.findSchemaByContent(SubjectsResourceImpl_Subclass.zig:452)
	at jdk.internal.reflect.GeneratedMethodAccessor93.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:249)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:35)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:229)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchToPath(ServletInitialHandler.java:164)
	at io.undertow.servlet.spec.RequestDispatcherImpl.forwardImpl(RequestDispatcherImpl.java:227)
	at io.undertow.servlet.spec.RequestDispatcherImpl.forwardImplSetup(RequestDispatcherImpl.java:149)
	at io.undertow.servlet.spec.RequestDispatcherImpl.forward(RequestDispatcherImpl.java:111)
	at io.apicurio.registry.rest.RegistryApplicationServletFilter.doFilter(RegistryApplicationServletFilter.java:153)
	at io.apicurio.registry.rest.RegistryApplicationServletFilter_ClientProxy.doFilter(RegistryApplicationServletFilter_ClientProxy.zig:196)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:133)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:65)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:247)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:56)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:111)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:108)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$9$1.call(UndertowDeploymentRecorder.java:593)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:227)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleRequest(ServletInitialHandler.java:152)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$1.handleRequest(UndertowDeploymentRecorder.java:119)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:290)
	at io.undertow.server.DefaultExchangeHandler.handle(DefaultExchangeHandler.java:18)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$5$1.run(UndertowDeploymentRecorder.java:415)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.apicurio.registry.storage.impl.sql.jdb.RuntimeSqlException: SQL error: Expected only one result but got multiple.
	at io.apicurio.registry.storage.impl.sql.jdb.MappedQueryImpl.findOne(MappedQueryImpl.java:106)
	at io.apicurio.registry.storage.impl.sql.AbstractSqlRegistryStorage.lambda$getArtifactVersionMetaData$45(AbstractSqlRegistryStorage.java:1247)
	at io.apicurio.registry.storage.impl.sql.HandleFactory.withHandle(HandleFactory.java:44)
	at io.apicurio.registry.storage.impl.sql.HandleFactory_ClientProxy.withHandle(HandleFactory_ClientProxy.zig:158)
	at io.apicurio.registry.storage.impl.sql.AbstractSqlRegistryStorage.getArtifactVersionMetaData(AbstractSqlRegistryStorage.java:1236)
	... 126 more
```

--- a/app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java
+++ b/app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java
@@ -1281,7 +1281,7 @@ public abstract class AbstractSqlRegistryStorage extends AbstractRegistryStorage
                         .bind(2, artifactId)
                         .bind(3, hash)
                         .map(ArtifactVersionMetaDataDtoMapper.instance)
-                        .findOne();
+                        .findLast();
                 return res.orElseThrow(() -> new ArtifactNotFoundException(groupId, artifactId));
             });
         } catch (ArtifactNotFoundException e) {

--- a/app/src/main/java/io/apicurio/registry/storage/impl/sql/CommonSqlStatements.java
+++ b/app/src/main/java/io/apicurio/registry/storage/impl/sql/CommonSqlStatements.java
@@ -207,7 +207,7 @@ public abstract class CommonSqlStatements implements SqlStatements {
         return "SELECT v.*, a.type FROM versions v "
                 + "JOIN content c ON v.contentId = c.contentId AND v.tenantId = c.tenantId "
                 + "JOIN artifacts a ON v.tenantId = a.tenantId AND v.groupId = a.groupId AND v.artifactId = a.artifactId "
-                + "WHERE v.tenantId = ? AND v.groupId = ? AND v.artifactId = ? AND c.contentHash = ?";
+                + "WHERE v.tenantId = ? AND v.groupId = ? AND v.artifactId = ? AND c.contentHash = ? ORDER BY v.globalId";
     }
 
     @Override
@@ -226,7 +226,7 @@ public abstract class CommonSqlStatements implements SqlStatements {
         return "SELECT v.*, a.type FROM versions v "
                 + "JOIN content c ON v.contentId = c.contentId AND v.tenantId = c.tenantId "
                 + "JOIN artifacts a ON v.tenantId = a.tenantId AND v.groupId = a.groupId AND v.artifactId = a.artifactId "
-                + "WHERE v.tenantId = ? AND v.groupId = ? AND v.artifactId = ? AND c.canonicalHash = ?";
+                + "WHERE v.tenantId = ? AND v.groupId = ? AND v.artifactId = ? AND c.canonicalHash = ? ORDER BY v.globalId";
     }
 
     /**

--- a/app/src/main/java/io/apicurio/registry/storage/impl/sql/jdb/MappedQuery.java
+++ b/app/src/main/java/io/apicurio/registry/storage/impl/sql/jdb/MappedQuery.java
@@ -33,6 +33,8 @@ public interface MappedQuery<R> {
 
     public Optional<R> findFirst();
 
+    public Optional<R> findLast();
+
     public List<R> list();
 
     public Stream<R> stream();

--- a/app/src/main/java/io/apicurio/registry/storage/impl/sql/jdb/MappedQueryImpl.java
+++ b/app/src/main/java/io/apicurio/registry/storage/impl/sql/jdb/MappedQueryImpl.java
@@ -137,6 +137,27 @@ public class MappedQueryImpl<T> implements MappedQuery<T>, Closeable {
     }
 
     /**
+     * @see io.apicurio.registry.storage.impl.sql.jdb.MappedQuery#findLast()
+     */
+    @Override
+    public Optional<T> findLast() {
+        Optional<T> rval = null;
+        try {
+            while (this.resultSet.next()) {
+                rval = Optional.of(this.mapper.map(resultSet));
+            }
+            if (rval == null) {
+                rval = Optional.empty();
+            }
+        } catch (SQLException e) {
+            throw new RuntimeSqlException(e);
+        } finally {
+            close();
+        }
+        return rval;
+    }
+
+    /**
      * @see io.apicurio.registry.storage.impl.sql.jdb.MappedQuery#list()
      */
     @Override

--- a/app/src/test/java/io/apicurio/registry/rbac/RegistryClientTest.java
+++ b/app/src/test/java/io/apicurio/registry/rbac/RegistryClientTest.java
@@ -19,6 +19,7 @@ package io.apicurio.registry.rbac;
 import static io.apicurio.registry.utils.tests.TestUtils.retry;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -1507,6 +1508,29 @@ public class RegistryClientTest extends AbstractResourceTestBase {
         } finally {
             mock.stop();
         }
+    }
+
+    @Test
+    public void testGetArtifactVersionByContent_DuplicateContent() throws Exception {
+        //Preparation
+        final String groupId = "testGetArtifactVersionByContent_DuplicateContent";
+        final String artifactId = generateArtifactId();
+
+        final ArtifactMetaData v1md = createArtifact(groupId, artifactId);
+
+        final InputStream v2stream = IoUtil.toStream(ARTIFACT_CONTENT.getBytes(StandardCharsets.UTF_8));
+        final VersionMetaData v2md = clientV2.createArtifactVersion(groupId, artifactId, null, v2stream);
+
+        //Execution
+        final VersionMetaData vmd = clientV2.getArtifactVersionMetaDataByContent(groupId, artifactId, IoUtil.toStream(ARTIFACT_CONTENT.getBytes()));
+
+        //Assertions
+        assertNotEquals(v1md.getGlobalId(), v2md.getGlobalId());
+        assertEquals(v1md.getContentId(), v2md.getContentId());
+
+        assertEquals(v2md.getGlobalId(), vmd.getGlobalId());
+        assertEquals(v2md.getId(), vmd.getId());
+        assertEquals(v2md.getContentId(), vmd.getContentId());
     }
 
 }

